### PR TITLE
Delete Iceberg table data on DROP

### LINKED ISSUE #5616: Iceberg DROP table not removing data from s3
Hi, 
I am using presto _(version 343)_ with Iceberg connector and found issue:

Tables which was created as `CREATE TABLE ...` or `CREATE TABLE AS SELECT...` then dropped - it will remove data only from Hive Metastore but **not** from s3.
Then if we try to create same table again - we will have `table already exists` exception

**Example:**

```sql
CREATE TABLE iceberg.test_table (id varchar(25));
INSERT INTO iceberg.test_table values ('1');
DROP TABLE iceberg.test_table

-- Table removed from Hive metastore, but files still exists in s3
```

Note:
In connector config we have `hive.metastore.thrift.delete-files-on-drop=true`

**Issue investigation results**
1. **CREATE TABLE** - when we create table (see [HiveTableOperations](https://github.com/prestosql/presto/blob/6e43e24194f7d3cb61d5222dd7c9ccafdc4f4873/presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/HiveTableOperations.java#L185)), we have hardcoded TableType=EXTERNAL_TABLE and parameter (EXTERNAL = TRUE)

```java
             Table.Builder builder = Table.builder()
                        ...
                        .setTableType(TableType.EXTERNAL_TABLE.name())
                        ...
                        .setParameter("EXTERNAL", "TRUE")
                        ...
```
2. **DROP TABLE** - we have 2 places which responsible for data deletion from s3
2.1 First one is Hive Metastore - it will delete files from **EXTERNAL** table only if we passed `external.table.purge=TRUE` parameter to it 
2.2. Second place is [ThriftHiveMetastore](https://github.com/prestosql/presto/blob/47dbba4140f474b3179e54176bfb12523f76ca57/presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftHiveMetastore.java#L1093), but as we have hardcoded TableType=EXTERNAL_TABLE - this code will be skipped
```java
  if (deleteFilesOnDrop && deleteData && isManagedTable(table) && !isNullOrEmpty(tableLocation)) {
                                deleteDirRecursive(hdfsContext, hdfsEnvironment, new Path(tableLocation));
                            }
#isManagedTable(table) -> false, because on table create we set tableType = EXTERNAL_TABLE
```

As a result data from s3 not deleted

### PR DESCRIPTION
**Pull request by findinpath**
## Description

Delegate the responsibility of deleting the data files
of the table to Iceberg because the data files
of the Iceberg table may be located in different locations.

The DROP statement deletes the Iceberg table data irrespective
of the fact that the `location` parameter has been specified
explicitly on the table, or it was implicitly derived from
the schema location.

It is important to note that the DROP functionality can be currently 
performed only for Iceberg tables that do not contain path overrides
(`write.object-storage.path`, `write.folder-storage.path`, `write.metadata.path`).

## General information

> Is this change a fix, improvement, new feature, refactoring, or other?

This is a fix.
Previously the data of the Iceberg table was not removed by HMS because Iceberg tables are marked implicitly as EXTERNAL, reason why the data of the table is not deleted by HMS on DROP.

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

This change relates solely to the Iceberg connector.

> How would you describe this change to a non-technical end user or system administrator?

This feature brings the ability to delete the data files while performing a DROP statements.

## Related issues, pull requests, and links

Fixes #5616
Related PR: [#6063](https://github.com/trinodb/trino/pull/6063)
Depends on PR: [#11032](https://github.com/trinodb/trino/pull/11032)

## Documentation

( ) No documentation is needed.
(x) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Iceberg
* Delete data files on DROP for managed tables. ({issue}`5616`)
```

### CODE REVIEW

**Comment by findinpath** [comment]
@mosabua  can you please indicate whether we should document explicitly in the documentation of the connector the `DROP` functionality?
Currently there are no details provided on how `DROP` is supposed to work for managed/external Iceberg tables https://trino.io/docs/current/connector/iceberg.html

I had a look over https://docs.starburst.io/latest/connector/starburst-delta-lake.html which implements as well this concept and haven't seen any reference there for `DROP` either.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java [hunk 1]
```diff
@@ -323,7 +318,7 @@ else if (table.getTableType().equals(EXTERNAL_TABLE.name())) {
                 throw new TrinoException(HIVE_METASTORE_ERROR, "Could not validate external location", e);
             }
         }
-        else if (!table.getTableType().equals(MATERIALIZED_VIEW.name())) {
+        else if (!(table.getTableType().equals(MANAGED_TABLE.name()) || table.getTableType().equals(MATERIALIZED_VIEW.name()))) {
```
Comment (jirassimok) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java:321]: I'd reformat this check:
```java
else if (!table.getTableType.equals(MANAGED_TABLE.name())
        && !table.getTableType.equals(MATERIALIZED_VIEW.name())) {
```
I think `!a && !b` is a bit clearer than `!(a || b)` when it comes to excluding multiple values, and the split line lets it fit on a screen at a reasonable width.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/TableMetadata.java [hunk 2]
```diff
@@ -50,6 +50,7 @@
     private final Map<String, String> serdeParameters;
 
     private final Optional<String> externalLocation;
+    private final Optional<String> location;
```
Comment (jirassimok) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/TableMetadata.java:53]: Why is this change necessary?

(And should it be in a separate commit?)

**Review by jirassimok — COMMENTED** [review]
I think the first commit is a pretty good change, though I'd reword it to clarify that it only affects `FileHiveMetastore`. Something like "allow alternate table data locations in FileHiveMetastore."

I didn't look at the second commit in detail, but I think the commit message should note that it's basically the same way `HiveMetadata` handles the same information.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/TableMetadata.java (hunk 2 above)
Comment (findinpath) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/TableMetadata.java:53]: This is needed to expose the correct `table.storage.location` . Before this change, only the external tables were allowed to expose a different location than the table name within the schema directory.

**Comment by alexjo2144** [comment]
I was just taking a look through the Iceberg API utilities, did you see https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/CatalogUtil.java#L78 ?

Looks like it may be a safer way of dropping table files

**Comment by findinpath** [comment]
> I was just taking a look through the Iceberg API utilities, did you see https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/CatalogUtil.java#L78 ? 

@alexjo2144 yes I did see also this utility function and to be honest, I was planning to delegate the deletion logic to this method. However, in the current state of Trino, there are tables being dropped only when they don't have any overridden location properties. So, **in the current stage of the Iceberg connector implementation**, the deletion of the table data can be still delegated to the Hive Metastore.

**Comment by alexjo2144** [comment]
I think the main advantage to using that utility instead of delegating the delete to Hive is consistency between tables made my Spark and tables made by Trino. 

If I follow the same steps:
- Create a schema with a location
- Create a table using the default location

In either Spark or Trino, and then drop the tables from Trino, I would expect it to do the same thing in either case, but that's not true here since we're relying on the managed/external flag being set on table creation and Spark will always set it to be external.

Does that make sense?

**Comment by findinpath** [comment]
@alexjo2144  this goes rather in the direction of interpreting all tables as EXTERNAL (the current status quo before this PR).
If I understand correctly, Trino would take then the responsibility of removing all the Iceberg tables on DROP TABLE.

**Comment by jirassimok** [comment]
> interpreting all tables as EXTERNAL (the current status quo before this PR)

I don't think this is the case. I believe a table should only be considered external if it has the `external_location` property.

**Comment by alexjo2144** [comment]
> I don't think this is the case. I believe a table is only supposed to be considered external if it has the external_location property.

In the Hive connector, yes, but Iceberg does not really have a well defined managed/external definition yet.

> this goes rather in the direction of interpreting all tables as EXTERNAL (the current status quo before this PR).
If I understand correctly, Trino would take then the responsibility of removing all the Iceberg tables on DROP TABLE.

That's correct, personally I think this would be a better approach.

**Comment by findinpath** [comment]
I went ahead and looked through the iceberg's implementation for spark `TableCatalog` and see in `org.apache.iceberg.hive.HiveCatalog#dropTable`
https://github.com/apache/iceberg/blob/e1c801653db05f5ac2fbaecf36be4320806c27b2/hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java#L163-L172

```
    try {
      clients.run(client -> {
        client.dropTable(database, identifier.name(),
            false /* do not delete data */,
            false /* throw NoSuchObjectException if the table doesn't exist */);
        return null;
      });

      if (purge && lastMetadata != null) {
        CatalogUtil.dropTableData(ops.io(), lastMetadata);
      }
```

As it can be seen, Spark does delegate to `CatalogUtil` the responsibility of deleting the table data in a file-by-file fashion.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java [hunk 3]
```diff
@@ -524,12 +524,10 @@ public synchronized void dropTable(String databaseName, String tableName, boolea
 
         Path tableMetadataDirectory = getTableMetadataDirectory(databaseName, tableName);
 
-        // It is safe to delete the whole meta directory for external tables and views
-        if (!table.getTableType().equals(MANAGED_TABLE.name()) || deleteData) {
```
Comment (findinpath) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java:528]: In the context of Iceberg tables, it is "not safe" anymore to delete the directory corresponding to the table.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java [hunk 4]
```diff
@@ -527,6 +522,12 @@ public synchronized void dropTable(String databaseName, String tableName, boolea
         // It is safe to delete the whole meta directory for external tables and views
         if (!table.getTableType().equals(MANAGED_TABLE.name()) || deleteData) {
             deleteDirectoryAndSchema(TABLE, tableMetadataDirectory);
+            if (table.getTableType().equals(MANAGED_TABLE.name())) {
+                Path tableDataLocation = new Path(table.getStorage().getLocation());
+                if (!tableMetadataDirectory.equals(tableDataLocation)) {
+                    deleteTableData(tableDataLocation);
+                }
+            }
         }
```
[the remark this answers is not in the document]
Comment (findinpath) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java]: Outdated request in the context of leaving Iceberg tables as EXTERNAL

**Comment by lzeiming** [comment]
> I think the main advantage to using that utility instead of delegating the delete to Hive is consistency between tables made my Spark and tables made by Trino.
> 
> If I follow the same steps:
> 
> * Create a schema with a location
> * Create a table using the default location
> 
> In either Spark or Trino, and then drop the tables from Trino, I would expect it to do the same thing in either case, but that's not true here since we're relying on the managed/external flag being set on table creation and Spark will always set it to be external.
> 
> Does that make sense?

I think so. Renaming a managed table may cause Hive metastore to do table location path rename on the filesystem. I think it is problematic for iceberg

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 5]
```diff
@@ -1098,6 +1098,9 @@ private void testCreateTableLikeForFormat(IcebergFileFormat otherFormat)
         File tempDir = getDistributedQueryRunner().getCoordinator().getBaseDataDir().toFile();
         String tempDirPath = tempDir.toURI().toASCIIString() + randomTableSuffix();
 
+        // LIKE source INCLUDING PROPERTIES copies all the properties of the source table, including the `location`.
```
Comment (alexjo2144) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:1101]: Instead, can you turn on useUniqueTableLocations?

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 6]
```diff
@@ -3373,4 +3370,16 @@ public void testTargetMaxFileSize()
                 // so just to be safe we check if it is not much bigger
                 .forEach(row -> assertThat((Long) row.getField(0)).isBetween(1L, maxSize.toBytes() * 3));
     }
+
+    @Test
+    public void testDroppingIcebergAndCreatingANewTableWithTheSameNameShouldBePossible()
+    {
+        assertUpdate("CREATE TABLE test_iceberg_recreate (x1 BIGINT)");
+        assertThat(query("SELECT count(x1) FROM test_iceberg_recreate")).matches("VALUES BIGINT '0'");
```
Comment (alexjo2144) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:3378]: Would be a better test if you insert something here so that after the drop there's a different result

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoHiveCatalog.java [hunk 7]
```diff
@@ -298,13 +303,51 @@ public Transaction newCreateTableTransaction(
     public void dropTable(ConnectorSession session, SchemaTableName schemaTableName, boolean purgeData)
     {
         // TODO: support path override in Iceberg table creation: https://github.com/trinodb/trino/issues/8861
-        Table table = loadTable(session, schemaTableName);
+        TableMetadata metadata = tableMetadataCache.computeIfAbsent(
+                schemaTableName,
+                ignore -> ((BaseTable) loadIcebergTable(metastore, tableOperationsProvider, session, schemaTableName)).operations().current());
+
+        Table table = getIcebergTableWithMetadata(metastore, tableOperationsProvider, session, schemaTableName, metadata);
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoHiveCatalog.java:310]: fix code layout after inline

```suggestion
        Table table = getIcebergTableWithMetadata(metastore, tableOperationsProvider, session, schemaTableName, metadata);

```

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoHiveCatalog.java [hunk 8]
```diff
@@ -298,13 +303,51 @@ public Transaction newCreateTableTransaction(
...
         }
-        metastore.dropTable(schemaTableName.getSchemaName(), schemaTableName.getTableName(), purgeData);
+        io.trino.plugin.hive.metastore.Table metastoreTable = metastore.getTable(schemaTableName.getSchemaName(), schemaTableName.getTableName())
+                .orElseThrow(() -> new TableNotFoundException(schemaTableName));
+
+        metastore.dropTable(
+                schemaTableName.getSchemaName(),
+                schemaTableName.getTableName(),
+                false /* do not delete data */);
+        if (purgeData) {
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoHiveCatalog.java:323]: `purgeData` is always true (https://github.com/trinodb/trino/pull/11167), so:

```suggestion
        verify(purgeData); // purgeData is always true
```

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoHiveCatalog.java [hunk 9]
```diff
@@ -298,13 +303,51 @@ public Transaction newCreateTableTransaction(
...
+            // Use the Iceberg routine for dropping the table data because the data files
+            // of the Iceberg table may be located in different locations
+            dropTableData(table.io(), metadata);
+
+            //TODO remove the following call once the upgrade to Iceberg 0.13.x is done https://github.com/trinodb/trino/pull/11032 . See https://github.com/apache/iceberg/pull/3622
+            Tasks.foreach(Iterables.transform(metadata.previousFiles(), TableMetadata.MetadataLogEntry::file))
+                    .noRetry().suppressFailureWhenFinished()
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoHiveCatalog.java:330]: why (both?)

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java [hunk 10]
```diff
@@ -524,12 +524,10 @@ public synchronized void dropTable(String databaseName, String tableName, boolea
 
         Path tableMetadataDirectory = getTableMetadataDirectory(databaseName, tableName);
 
-        // It is safe to delete the whole meta directory for external tables and views
-        if (!table.getTableType().equals(MANAGED_TABLE.name()) || deleteData) {
+        if (deleteData) {
```
[the remark this answers is not in the document]
Comment (findinpath) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java:527]: if the FileHiveMetastore deletes the table directory, even though `deleteData` was set to `false`, the `dropTableData` routine from TrinoHiveCatalog (which happens after the `metastore.dropTable` call) will not work anymore.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 11]
```diff
@@ -1098,6 +1098,9 @@ private void testCreateTableLikeForFormat(IcebergFileFormat otherFormat)
         File tempDir = getDistributedQueryRunner().getCoordinator().getBaseDataDir().toFile();
         String tempDirPath = tempDir.toURI().toASCIIString() + randomTableSuffix();
 
+        // LIKE source INCLUDING PROPERTIES copies all the properties of the source table, including the `location`.
+        // For this reason the source and the copied table will share the same directory.
+        // This test does not drop intentionally the created tables to avoid affecting the source table or the information_schema.
```
[the remark this answers is not in the document]
Comment (findinpath) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:1103]: At the moment, dropping one of the tables breaks the drop procedure for the other tables.
Without specific logic to avoid mixing the content of the tables in the same directory I don't know how to perform the `DROP` statements.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkDropTableCompatibility.java [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+    private static List<String> getDataFilePaths(String icebergTableName)
+    {
+        List<String> filePaths = onTrino().executeQuery(format("SELECT file_path FROM %s.%s.\"%s$files\"", TRINO_CATALOG, TEST_SCHEMA_NAME, icebergTableName)).column(1);
+        return filePaths.stream().map(TestIcebergSparkDropTableCompatibility::getPath).collect(Collectors.toList());
+    }
+
+    private static String getPath(String uri)
+    {
+        try {
+            return new URI(uri).getPath();
```
[the remark this answers is not in the document]
Comment (findinpath) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkDropTableCompatibility.java:104]: > path of an uri (discarding scheme and authority) seems wrong

The `path` is then given to the `hdfsClient` which makes use of the scheme and authority behind the scenes.

The `hdfs` bash utility makes use of paths instead of URIs

```
hdfs dfs -ls -R /user/hive/warehouse
```


Not necessarily as an argument, I found two instances of this usage pattern:

- `io.trino.tests.product.TestImpersonation#getTableLocation`
- `io.trino.tests.product.hive.util.TableLocationUtils#getTablePath(java.lang.String, int)`

[the remark this answers is not in the document]
Comment (findepi) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkDropTableCompatibility.java:104]: > Not necessarily as an argument, I found two instances of this usage pattern:

Can we try better than that, since now we realized it's not necessarily what we should be doing?

note that `new URI(uri).getPath();`  transformation is simply lossy:

```
hdfs://host/path -> path
s3://bucket/path -> path
```

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoHiveCatalog.java [hunk 13]
```diff
@@ -298,13 +303,51 @@ public Transaction newCreateTableTransaction(
...
+        metastore.dropTable(
+                schemaTableName.getSchemaName(),
+                schemaTableName.getTableName(),
+                false /* do not delete data */);
+        if (purgeData) {
+            // Use the Iceberg routine for dropping the table data because the data files
+            // of the Iceberg table may be located in different locations
+            dropTableData(table.io(), metadata);
+
+            //TODO remove the following call once the upgrade to Iceberg 0.13.x is done https://github.com/trinodb/trino/pull/11032 . See https://github.com/apache/iceberg/pull/3622
```
[the remark this answers is not in the document]
Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoHiveCatalog.java]: that PR got merged. let's make sure this is updated here then

**Comment by findinpath** [comment]
Rebasing the PR on master to make use of the update to iceberg library to the version `0.13.1`.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkDropTableCompatibility.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (findinpath) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkDropTableCompatibility.java]: > if we need to strip off a known prefix (eg hdfs://....), this can be done with uri.replaceFirst("^\\Q....")`,

I fail to see at the moment what the appropriate regex would be in such a case. Using a host/ip(&port) regex seems fairly complicated for readability.

[the remark this answers is not in the document]
Comment (findepi) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkDropTableCompatibility.java]: is the hostname always `hadoop-master` in product tests?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoHiveCatalog.java (hunk 7 above)
Comment (findinpath) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoHiveCatalog.java]: @findepi the code snippets are the same. Can you please rewrite the snippet suggestion?

Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoHiveCatalog.java]: @findinpath the difference is in where the empty line is positioned.
i see the github UI is not displaying this clearly

Comment (findinpath) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoHiveCatalog.java]: Applied cosmetic changes on the method code.
I hope I did understand correctly what you meant.

**Review by findepi — APPROVED** [review]
